outlines/fsm/json_schema.py

```diff
@@ -41,7 +50,211 @@
+                for sub_min, sub_max in get_subranges(minimum, maximum)
+            ]
+        )
+
+    if explicit_zero:
+        pattern = rf"(0|({pattern}))"
+
+    return pattern
+
+
+def get_int_pattern(minimum=None, maximum=None, safe_subset: bool = False):
+    """
+    This function generates a regex pattern that matches integers from `minimum` to `maximum`,
+    inclusive. It handles negative ranges, positive ranges, zero, and ranges that span both negative
+    and positive numbers.
+
+    If no bounds are specified, it defaults to matching all integers. The `safe_subset` parameter
+    can be used to limit the range to safe integer values (e.g., to avoid excessively large numbers).
+
+    Args:
+        minimum (int, (+/-)inf, optional): The lower bound of the integer range. Defaults to negative infinity.
+        maximum (int, (+/-)inf, optional): The upper bound of the integer range. Defaults to positive infinity.
+        safe_subset (bool, optional): If True, uses SAFE_INT_MIN and SAFE_INT_MAX as default bounds.
+
+    Returns:
+        str: A regex pattern string that matches all integers in the specified range.
+    """
+    # handle safe subset of range
+    if minimum is None:
+        minimum = SAFE_INT_MIN if safe_subset else -float("inf")
+    if maximum is None:
+        maximum = SAFE_INT_MAX if safe_subset else float("inf")
+
+    if (minimum, maximum) == (-float("inf"), float("inf")):
+        return INTEGER
+
+    assert minimum <= maximum
+
+    if minimum == maximum == 0:
+        pattern = "0"
+    elif minimum >= 0 and maximum >= 0:
+        pattern = get_positive_int_range_pattern(minimum, maximum)
+    elif minimum < 0 and maximum <= 0:
+        # entirely negative range: prefix with `-` and calculate abs of range
+        abs_pattern = get_positive_int_range_pattern(max(abs(maximum), 1), abs(minimum))
+        pattern = rf"-({abs_pattern})"
+        if maximum == 0:
+            pattern = rf"0|({pattern})"
+    else:  # minimum < 0 and maximum > 0:
+        # positive component of range | negative component
+        minimum_pattern = get_positive_int_range_pattern(1, abs(minimum))
+        maximum_pattern = get_positive_int_range_pattern(0, maximum)
+        pattern = rf"(-({minimum_pattern}))|({maximum_pattern})"
+
+    return rf"({pattern})"
+
+
+def get_str_pattern(min_length=None, max_length=None, safe_subset=False):
+    if min_length is None and max_length is None and not safe_subset:
+        return STRING
+    elif min_length and max_length and int(max_length) < int(min_length):
+        raise ValueError("maxLength must be greater than or equal to minLength")
+    elif min_length and min_length < 0:
+        raise ValueError("minLength must be greater than or equal to 0")
+
+    min_length = min_length or ""
+    if max_length is None:
+        max_length = str(SAFE_STR_MAX_LEN) if safe_subset else ""
+
+    return f'"{STRING_INNER}{{{min_length},{max_length}}}"'
+
+
+def build_regex_from_schema(
```

Comment:
Probably should be at the top. We might also consider separating into multiple modules so it's easier to read

Reply:
Yes, I generally like to move helper functions to the end of the file.